python/ray/data/dataset.py

```diff
@@ -521,13 +521,23 @@ def transform(
             *fn_args,
             **fn_kwargs,
         ) -> Iterable[Block]:
+
+            # Fetch next available input blocks for the batch.
+            # Only fetch required number of blocks but not all blocks.
+            # This avoids over-buffering in current node's memory.
+            def _fetch_next_blocks(batcher: Batcher, batch_size: Optional[int]):
+                for block in blocks:
+                    batcher.add(block)
```

Comment:
Should this keep track of the block index? Otherwise won't this always start from the first block?

Reply:
Sorry why we need to keep track the block index? `blocks` is an `Iterable[Block]`, so it will get the next element from iterator.

Reply:
Right, `blocks` is an `Iterable` but not an `Iterator` - each time `_fetch_next_blocks` is called wouldn't it start iterating from the beginning?

Reply:
@matthewdeng - ah good catch, thanks. I was only testing with generator, but didn't test `List` for blocks. Fixed.